cli/main.c

```diff
@@ -90,6 +90,23 @@ int switchtec_handler(const char *optarg, void *value_addr,
 	return 0;
 }
 
+int mfg_handler(const char *optarg, void *value_addr,
+		const struct argconfig_options *opt)
+{
+	int fd;
+
+	fd = open(optarg, O_RDWR);
+	if (fd >= 0 && isatty(fd)) {
```

Comment:
Yeah, no I really don't like this... For starters it's a terrible way to tell if you are working with a UART, secondly, it does nothing for library users.

Here's a rough untested diff that would be better. (Preferably done in a couple commits):

```Diff

diff --git a/inc/switchtec/errors.h b/inc/switchtec/errors.h
index c020bfcaf497..d37361a96331 100755
--- a/inc/switchtec/errors.h
+++ b/inc/switchtec/errors.h
@@ -59,6 +59,8 @@ enum {
 	ERR_EVENT_INVALID = 0x100003,
 	ERR_RST_RULE_FAILED = 0x100005,
 	ERR_ACCESS_REFUSED = 0xFFFF0001,
+
+	ERR_UART_NOT_SUPPORTED = 0x10000001,
 };
 
 #endif
diff --git a/lib/mfg.c b/lib/mfg.c
index 8b0420377059..90df281745bc 100755
--- a/lib/mfg.c
+++ b/lib/mfg.c
@@ -89,6 +89,19 @@ static void RSA_get0_key(const RSA *r, const BIGNUM **n,
 }
 #endif
 
+static int switchtec_mfg_cmd(struct switchtec_dev *dev, uint32_t cmd,
+			     const void *payload, size_t payload_len,
+			     void *resp, size_t resp_len);
+{
+	if (dev->ops->flags & SWITCHTEC_OPS_FLAG_NO_MFG) {
+		errno = ERR_UART_NOT_SUPPORTED | SWITCHTEC_ERRNO_MRPC_FLAG_BIT;
+		return -1;
+	}
+
+	return switchtec_cmd(dev, cmd, payload, payload_len,
+			     resp, resp_len);
+}
+
 /**
  * @brief Get serial number and security version
  * @param[in]  dev	Switchtec device handle
@@ -100,8 +113,8 @@ int switchtec_sn_ver_get(struct switchtec_dev *dev,
 {
 	int ret;
 
-	ret = switchtec_cmd(dev, MRPC_SN_VER_GET, NULL, 0, info,
-			    sizeof(struct switchtec_sn_ver_info));
+	ret = switchtec_mfg_cmd(dev, MRPC_SN_VER_GET, NULL, 0, info,
+				sizeof(struct switchtec_sn_ver_info));
 	if (ret)
 		return ret;
 
@@ -137,8 +150,8 @@ int switchtec_security_config_get(struct switchtec_dev *dev,
 		uint8_t  rsvd4[32];
 	} reply;
 
-	ret = switchtec_cmd(dev, MRPC_SECURITY_CONFIG_GET, NULL, 0,
-			    &reply, sizeof(reply));
+	ret = switchtec_mfg_cmd(dev, MRPC_SECURITY_CONFIG_GET, NULL, 0,
+				&reply, sizeof(reply));
 	if (ret)
 		return ret;
 
@@ -200,8 +213,8 @@ int switchtec_mailbox_to_file(struct switchtec_dev *dev, int fd)
 	} reply;
 
 	do {
-		ret = switchtec_cmd(dev, MRPC_MAILBOX_GET, &num_to_read,
-				    sizeof(int), &reply,  sizeof(reply));
+		ret = switchtec_mfg_cmd(dev, MRPC_MAILBOX_GET, &num_to_read,
+					sizeof(int), &reply,  sizeof(reply));
 		if (ret)
 			return ret;
 
@@ -263,8 +276,8 @@ int switchtec_security_config_set(struct switchtec_dev *dev,
 
 	sd.pub_key_exponent = htole32(setting->public_key_exponent);
 
-	ret = switchtec_cmd(dev, MRPC_SECURITY_CONFIG_SET, &sd, sizeof(sd),
-			    NULL, 0);
+	ret = switchtec_mfg_cmd(dev, MRPC_SECURITY_CONFIG_SET, &sd, sizeof(sd),
+				NULL, 0);
 	return ret;
 }
 
@@ -282,8 +295,8 @@ int switchtec_active_image_index_get(struct switchtec_dev *dev,
 		uint8_t index[SWITCHTEC_ACTV_IDX_MAX_ENTRIES];
 	} reply;
 
-	ret = switchtec_cmd(dev, MRPC_ACT_IMG_IDX_GET, NULL,
-			    0, &reply, sizeof(reply));
+	ret = switchtec_mfg_cmd(dev, MRPC_ACT_IMG_IDX_GET, NULL,
+				0, &reply, sizeof(reply));
 	if (ret)
 		return ret;
 
@@ -345,8 +358,8 @@ int switchtec_active_image_index_set(struct switchtec_dev *dev,
 
 	set.count = htole32(i);
 
-	ret = switchtec_cmd(dev, MRPC_ACT_IMG_IDX_SET, &set,
-			    sizeof(set), NULL, 0);
+	ret = switchtec_mfg_cmd(dev, MRPC_ACT_IMG_IDX_SET, &set,
+				sizeof(set), NULL, 0);
 	return ret;
 }
 
@@ -369,7 +382,7 @@ int switchtec_fw_exec(struct switchtec_dev *dev,
 	cmd.subcmd = MRPC_FW_TX_EXEC;
 	cmd.recovery_mode = recovery_mode;
 
-	return switchtec_cmd(dev, MRPC_FW_TX, &cmd, sizeof(cmd), NULL, 0);
+	return switchtec_mfg_cmd(dev, MRPC_FW_TX, &cmd, sizeof(cmd), NULL, 0);
 }
 
 /**
@@ -379,15 +392,15 @@ int switchtec_fw_exec(struct switchtec_dev *dev,
  *        calls to switchtec library functions, be sure to close
  *        this pointer and get a new one by calling switchtec_open().
  * 	  Also be sure to check the return value of switchtec_open()
- * 	  for error, as the device might not be immediately 
+ * 	  for error, as the device might not be immediately
  * 	  accessible after normal boot process.
  * @param[in]  dev	Switchtec device handle
  * @return 0 on success, error code on failure
  */
 int switchtec_boot_resume(struct switchtec_dev *dev)
 {
-	return switchtec_cmd(dev, MRPC_BOOTUP_RESUME, NULL, 0,
-			     NULL, 0);
+	return switchtec_mfg_cmd(dev, MRPC_BOOTUP_RESUME, NULL, 0,
+				 NULL, 0);
 }
 
 /**
@@ -407,8 +420,8 @@ int switchtec_secure_state_set(struct switchtec_dev *dev,
 	}
 	data = htole32(state);
 
-	return switchtec_cmd(dev, MRPC_SECURE_STATE_SET, &data, sizeof(data),
-			     NULL, 0);
+	return switchtec_mfg_cmd(dev, MRPC_SECURE_STATE_SET, &data,
+				 sizeof(data), NULL, 0);
 }
 
 static int dbg_unlock_send_pubkey(struct switchtec_dev *dev,
@@ -425,8 +438,8 @@ static int dbg_unlock_send_pubkey(struct switchtec_dev *dev,
 	memcpy(cmd.pub_key, public_key->pubkey, SWITCHTEC_PUB_KEY_LEN);
 	cmd.pub_key_exp = htole32(public_key->pubkey_exp);
 
-	return switchtec_cmd(dev, MRPC_DBG_UNLOCK, &cmd,
-			     sizeof(cmd), NULL, 0);
+	return switchtec_mfg_cmd(dev, MRPC_DBG_UNLOCK, &cmd,
+				 sizeof(cmd), NULL, 0);
 }
 
 /**
@@ -461,8 +474,8 @@ int switchtec_dbg_unlock(struct switchtec_dev *dev, uint32_t serial,
 	cmd.unlock_ver = htole32(ver_sec_unlock);
 	memcpy(cmd.signature, signature->signature, SWITCHTEC_SIG_LEN);
 
-	return switchtec_cmd(dev, MRPC_DBG_UNLOCK, &cmd,
-			     sizeof(cmd), NULL, 0);
+	return switchtec_mfg_cmd(dev, MRPC_DBG_UNLOCK, &cmd,
+				 sizeof(cmd), NULL, 0);
 }
 
 /**
@@ -498,8 +511,8 @@ int switchtec_dbg_unlock_version_update(struct switchtec_dev *dev,
 	cmd.unlock_ver = htole32(ver_sec_unlock);
 	memcpy(cmd.signature, signature->signature, SWITCHTEC_SIG_LEN);
 
-	return switchtec_cmd(dev, MRPC_DBG_UNLOCK, &cmd, sizeof(cmd),
-			     NULL, 0);
+	return switchtec_mfg_cmd(dev, MRPC_DBG_UNLOCK, &cmd, sizeof(cmd),
+				 NULL, 0);
 }
 
 /**
@@ -592,8 +605,8 @@ static int kmsk_set_send_pubkey(struct switchtec_dev *dev,
 	       SWITCHTEC_PUB_KEY_LEN);
 	cmd.pub_key_exponent = htole32(public_key->pubkey_exp);
 
-	return switchtec_cmd(dev, MRPC_KMSK_ENTRY_SET, &cmd,
-			     sizeof(cmd), NULL, 0);
+	return switchtec_mfg_cmd(dev, MRPC_KMSK_ENTRY_SET, &cmd,
+				 sizeof(cmd), NULL, 0);
 }
 
 static int kmsk_set_send_signature(struct switchtec_dev *dev,
@@ -609,8 +622,8 @@ static int kmsk_set_send_signature(struct switchtec_dev *dev,
 	memcpy(cmd.signature, signature->signature,
 	       SWITCHTEC_SIG_LEN);
 
-	return switchtec_cmd(dev, MRPC_KMSK_ENTRY_SET, &cmd,
-		    	     sizeof(cmd), NULL, 0);
+	return switchtec_mfg_cmd(dev, MRPC_KMSK_ENTRY_SET, &cmd,
+				 sizeof(cmd), NULL, 0);
 }
 
 static int kmsk_set_send_kmsk(struct switchtec_dev *dev,
@@ -627,8 +640,8 @@ static int kmsk_set_send_kmsk(struct switchtec_dev *dev,
 	cmd.num_entries = 1;
 	memcpy(cmd.kmsk, kmsk->kmsk, SWITCHTEC_KMSK_LEN);
 
-	return switchtec_cmd(dev, MRPC_KMSK_ENTRY_SET, &cmd, sizeof(cmd),
-			     NULL, 0);
+	return switchtec_mfg_cmd(dev, MRPC_KMSK_ENTRY_SET, &cmd, sizeof(cmd),
+				 NULL, 0);
 }
 
 /**
diff --git a/lib/platform/linux-uart.c b/lib/platform/linux-uart.c
index 81eb2999deda..695db2c69277 100755
--- a/lib/platform/linux-uart.c
+++ b/lib/platform/linux-uart.c
@@ -431,6 +431,7 @@ static ssize_t uart_write_from_gas(struct switchtec_dev *dev, int fd,
 }
 
 static const struct switchtec_ops uart_ops = {
+	.flags = SWITCHTEC_OPS_FLAG_NO_MFG,
 	.close = uart_close,
 	.gas_map = uart_gas_map,
 
@@ -536,4 +537,3 @@ err_free:
 }
 
 #endif
-
diff --git a/lib/switchtec.c b/lib/switchtec.c
index 395c4aa27924..be56c0048d74 100755
--- a/lib/switchtec.c
+++ b/lib/switchtec.c
@@ -537,7 +537,7 @@ const char *switchtec_strerror(void)
 	case ERR_EVENT_INVALID: 	msg = "Invalid Event"; break;
 	case ERR_RST_RULE_FAILED: 	msg = "Reset rule search failed"; break;
 	case ERR_ACCESS_REFUSED: 	msg = "Access Refused"; break;
-
+	case ERR_UART_NOT_SUPPORTED:    msg = "UART interface not supported for this command"; break;
 	default: break;
 	}
 
diff --git a/lib/switchtec_priv.h b/lib/switchtec_priv.h
index 391463d65802..87baa299ea0f 100755
--- a/lib/switchtec_priv.h
+++ b/lib/switchtec_priv.h
@@ -64,7 +64,12 @@ enum switchtec_fw_image_part_id_gen4 {
 	SWITCHTEC_FW_PART_ID_G4_SEEPROM = 0xFE,
 };
 
+enum switchtec_ops_flags {
+	SWITCHTEC_OPS_FLAG_NO_MFG = (1 << 0),
+};
+
 struct switchtec_ops {
+	int flags;
 	void (*close)(struct switchtec_dev *dev);
 	int (*get_device_id)(struct switchtec_dev *dev);
 	int (*get_fw_version)(struct switchtec_dev *dev, char *buf,
```

Reply:
Updated, please review.